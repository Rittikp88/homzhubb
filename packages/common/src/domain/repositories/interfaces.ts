import { ILastVisitedStep } from '@homzhub/common/src/domain/models/Asset';

// ENUMS
export enum LoginTypes {
  OTP = 'OTP_LOGIN',
  EMAIL = 'EMAIL_LOGIN',
  SOCIAL_MEDIA = 'SOCIAL_LOGIN',
}

export enum OtpActionTypes {
  SEND = 'SEND_OTP',
  VERIFY = 'VERIFY_OTP',
}

export enum SpaceAvailableTypes {
  BATHROOM = 'Bathroom',
  BEDROOM = 'Bedroom',
  TOTAL_FLOORS = 'Total Floors',
  FLOOR_NUMBER = 'Floor number',
  BALCONY = 'Balcony',
}

export enum VisitType {
  PHYSICAL = 'PHYSICAL',
  VIRTUAL = 'VIRTUAL',
  PROPERTY_VIEW = 'PROPERTY_VIEW',
}

// ENUMS - END

// USER AUTH - START
export interface ISignUpPayload {
  full_name: string;
  email: string;
  country_code: string;
  phone_number: string;
  password: string;
}

export interface ISocialSignUpPayload {
  otp: string;
  user_details: ISignUpPayload;
}

export interface IEmailLoginPayload {
  action: LoginTypes.EMAIL;
  payload: {
    email: string;
    password: string;
  };
}

export interface ILoginPayload {
  data: IEmailLoginPayload | IOtpLoginPayload;
  callback?: () => void;
}

export interface ISocialLogin {
  is_new_user?: boolean;
  access_token?: string;
  refresh_token?: string;
  user?: {
    full_name: string;
    email: string;
    country_code: string;
    phone_number: string;
  };
}

export interface ISocialLoginPayload {
  action: LoginTypes.SOCIAL_MEDIA;
  payload: {
    provider: string;
    id_token: string;
  };
}

export interface IOtpLoginPayload {
  action: LoginTypes.OTP;
  payload: {
    country_code: string;
    phone_number: string;
    otp: string;
  };
}

export interface ILoginFormData {
  email: string;
  password: string;
  country_code: string;
  phone_number: string;
}

export interface IForgotPasswordPayload {
  action: string;
  payload: {
    email?: string;
    token?: string | number;
    password?: string;
  };
}

export interface IResetPasswordData {
  email_exists: boolean;
  token: string;
}

export interface IUserExistsData {
  is_exists: boolean;
}

export interface IUserPayload {
  full_name: string;
  email: string;
  country_code: string;
  phone_number: string;
  access_token: string;
  refresh_token: string;
}

// USER AUTH - END

// OTP - START
export interface IOtpSendPayload {
  country_code: string;
  phone_number: string;
}

export interface IOtpVerifyPayload extends IOtpSendPayload {
  otp: string;
}

export interface IOtpVerify {
  action: OtpActionTypes;
  payload: IOtpSendPayload | IOtpVerifyPayload;
}

export interface IOtpVerifyResponse {
  otp_verify?: boolean;
  otp_sent?: boolean;
}

// OTP - END

// REFRESH TOKEN - START
export interface IRefreshToken {
  access_token: string;
  refresh_token: string;
}

export interface IRefreshTokenPayload {
  refresh_token: string;
}

// REFRESH TOKEN - END

// USER LOGOUT - START
export interface IUserLogoutPayload {
  action: string;
  payload: IRefreshTokenPayload;
}

// USER LOGOUT - END

// ASSET - START
export interface ISpaceAvailable {
  bedroom: number;
  bathroom: number;
  balcony: number;
  floorNumber: number;
  totalFloors: number;
  carpetArea: string;
  areaUnit: number;
}

export interface ISpaceAvailablePayload {
  space_type: string | number;
  name?: string;
  value?: number;
  count?: number;
  description?: string;
}

export interface ICreateAssetParams {
  project_name: string;
  unit_number: string;
  block_number: string;
  latitude: string | number;
  longitude: string | number;
  address: string;
  pin_code: string;
  city_name: string;
  state_name: string;
  country_name: string;
  country_iso2_code: string;
  asset_type: number;
  last_visited_step?: ILastVisitedStep;
}

export interface ICreateAssetResult {
  id: number;
}

export interface IUpdateAssetParams {
  project_name?: string;
  unit_number?: string;
  block_number?: string;
  latitude?: string | number;
  longitude?: string | number;
  address?: string;
  pin_code?: string;
  city_name?: string;
  state_name?: string;
  country_iso2_code?: string;
  country_name?: string;
  carpet_area?: number;
  carpet_area_unit?: number;
  floor_number?: number;
  total_floors?: number;
  asset_type?: number;
  last_visited_step?: ILastVisitedStep;
  spaces?: ISpaceAvailablePayload[];
  asset_highlights?: string[];
  is_gated?: boolean;
  power_backup?: boolean;
  corner_property?: boolean;
  all_day_access?: boolean;
  amenities?: number[];
  facing?: string;
  floor_type?: number;
  furnishing_description?: string;
  construction_year?: number;
  furnishing?: string;
}

// ASSET - END

// Property Search
export interface IPropertySearchPayload {
  asset_group: number;
  price__gte: number;
  price__lte: number;
  latitude: number;
  longitude: number;
  limit: number;
  offset: number;
  bathroom__gte: number;
  move_in_date__gte?: string;
  furnishing__in?: string;
  facing__in?: string;
  amenities__in?: string;
  rent_free_period?: number;
  is_verified?: boolean;
  agent_listed?: boolean;
  age__gte?: string;
  search_radius?: number;
  date_added?: string;
  asset_type__in?: string;
  bedroom__gte?: number;
  bedroom__in?: number;
  carpet_area__lt?: number;
  carpet_area__gt?: number;
  carpet_area_unit?: number;
}

// Lead
export interface ILeadPayload {
  propertyTermId: number;
  data: {
    spaces?: any;
    contact_person_type?: string;
    lead_type: string;
    message?: string;
    person_contacted?: number;
    preferred_contact_time?: any;
    is_wishlisted?: boolean;
    user_search?: null;
  };
}

export interface IGeneralLedgerPayload {
  transaction_date__lte: string;
  transaction_date__gte: string;
  transaction_date_group_by: string;
}

export interface IAddGeneralLedgerPayload {
  asset: number;
  entry_type: string;
  label: string;
  payer_name?: string;
  receiver_name?: string;
  amount: number;
  category: number;
  transaction_date: string;
  notes: string | null;
  attachment: number | null;
}

export interface ICreateLedgerResult {
  id: number;
}

export interface INotificationsPayload {
  limit?: number;
  offset?: number;
  q?: string;
  lease_listing_id?: number;
  sale_listing_id?: number;
}

export interface IDocumentPayload {
  attachment: number;
  lease_listing_id?: number;
  sale_listing_id?: number;
}

export interface ICreateDocumentPayload {
  propertyId: number;
  documentData: IDocumentPayload[];
}

export interface IUpcomingVisitPayload {
  start_date__gte: string;
  visit_type: VisitType;
  sale_listing_id: number | null;
  lease_listing_id: number | null;
}

export interface IScheduleVisitPayload {
  visit_type: VisitType;
  lead_type: number;
  start_date: string;
  end_date: string;
  sale_listing: number | null;
  lease_listing: number | null;
  comments?: string;
}

export interface IPropertyImagesPostPayload {
  attachment: number;
  is_cover_image: boolean;
}

export interface IMarkCoverImageAttachment {
  cover_updated: boolean;
}

export interface IGetSaleTermsParams {
  status: 'DRAFT';
}

export interface IOrderSummaryPayload {
  services?: number[];
  coins?: number;
  promo_code?: string;
}
