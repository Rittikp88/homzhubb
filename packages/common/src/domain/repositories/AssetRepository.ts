import { BootstrapAppService } from '@homzhub/common/src/services/BootstrapAppService';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { IPropertyDetailsData } from '@homzhub/common/src/domain/models/Property';
import { IApiClient } from '@homzhub/common/src/network/Interfaces';
import {
  ICreateAssetDetails,
  ICreateAssetResult,
  IUpdateAssetDetails,
} from '@homzhub/common/src/domain/repositories/interfaces';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { AssetReview } from '@homzhub/common/src/domain/models/AssetReview';
import {
  ILeaseTermDetails,
  ICreateLeaseTermDetails,
  IUpdateLeaseTermDetails,
} from '@homzhub/common/src/domain/models/LeaseTerms';
import {
  ICreateSaleTermDetails,
  ISaleDetails,
  IUpdateSaleTermDetails,
} from '@homzhub/common/src/domain/models/SaleTerms';
import {
  IMarkCoverImageAttachment,
  IPostVerificationDocuments,
  IPropertyImagesPostPayload,
  IPropertySelectedImages,
  IVerificationDocumentList,
  IVerificationTypes,
} from '@homzhub/common/src/domain/models/Service';

const ENDPOINTS = {
  createAsset: (): string => 'assets/',
  updateAsset: (id: number): string => `assets/${id}/`,
  getAssetById: (propertyId: number): string => `assets/${propertyId}/`,
  leaseTerms: (propertyId: number): string => `assets/${propertyId}/lease-terms/`,
  updateLeaseTerms: (propertyId: number, leaseTermId: number): string =>
    `assets/${propertyId}/lease-terms/${leaseTermId}/`,
  saleTerms: (propertyId: number): string => `assets/${propertyId}/sale-terms/`,
  updateSaleTerms: (propertyId: number, saleTermId: number): string => `assets/${propertyId}/sale-terms/${saleTermId}/`,
  existingVerificationDocuments: (propertyId: number): string => `assets/${propertyId}/verification-documents/`,
  deleteExistingVerificationDocuments: (propertyId: number, documentId: number): string =>
    `assets/${propertyId}/verification-documents/${documentId}/`,
  assetAttachments: (propertyId: number): string => `assets/${propertyId}/attachments/`,
  postVerificationDocuments: (propertyId: number): string => `assets/${propertyId}/verification-documents/`,
  markAttachmentAsCoverImage: (propertyId: number, attachmentId: number): string =>
    `assets/${propertyId}/attachments/${attachmentId}/cover-image`,
  getPropertyDetails: (): string => 'asset-groups/',
  deletePropertyAttachment: (attachmentId: number): string => `attachments/${attachmentId}`,
  assetIdentityDocuments: (): string => 'asset-identity-documents/',
  getVerificationDocumentDetails: (): string => 'verification-document-types/',
  getRatings: (id: number): string => `assets/${id}/ratings`,
  getLeaseListing: (propertyTermId: number): string => `lease-listings/${propertyTermId}`,
  getSaleListing: (propertyTermId: number): string => `sale-listings/${propertyTermId}`,
  getSimilarPropertiesForLease: (propertyTermId: number): string =>
    `lease-listings/${propertyTermId}/similar-properties/`,
  getSimilarPropertiesForSale: (propertyTermId: number): string =>
    `sale-listings/${propertyTermId}/similar-properties/`,
};

class AssetRepository {
  private apiClient: IApiClient;

  public constructor() {
    this.apiClient = BootstrapAppService.clientInstance;
  }

  public getAssetById = async (propertyId: number): Promise<Asset> => {
    const response = await this.apiClient.get(ENDPOINTS.getAssetById(propertyId));
    return ObjectMapper.deserialize(Asset, response);
  };

  public getDetails = async (): Promise<IPropertyDetailsData[]> => {
    return await this.apiClient.get(ENDPOINTS.getPropertyDetails());
  };

  public createAsset = async (assetDetails: ICreateAssetDetails): Promise<ICreateAssetResult> => {
    return await this.apiClient.post(ENDPOINTS.createAsset(), assetDetails);
  };

  public updateAsset = async (id: number, requestBody: IUpdateAssetDetails): Promise<void> => {
    return await this.apiClient.patch(ENDPOINTS.updateAsset(id), requestBody);
  };

  public getLeaseTerms = async (propertyId: number): Promise<ILeaseTermDetails[]> => {
    return await this.apiClient.get(ENDPOINTS.leaseTerms(propertyId));
  };

  public createLeaseTerms = async (
    propertyId: number,
    leaseTerms: ICreateLeaseTermDetails
  ): Promise<{ id: number }> => {
    return await this.apiClient.post(ENDPOINTS.leaseTerms(propertyId), leaseTerms);
  };

  public updateLeaseTerms = async (
    propertyId: number,
    leaseTermId: number,
    leaseTerms: IUpdateLeaseTermDetails
  ): Promise<void> => {
    return await this.apiClient.put(ENDPOINTS.updateLeaseTerms(propertyId, leaseTermId), leaseTerms);
  };

  public getSaleTerms = async (propertyId: number): Promise<ISaleDetails[]> => {
    return await this.apiClient.get(ENDPOINTS.saleTerms(propertyId));
  };

  public createSaleTerms = async (propertyId: number, saleTerms: ICreateSaleTermDetails): Promise<{ id: number }> => {
    return await this.apiClient.post(ENDPOINTS.saleTerms(propertyId), saleTerms);
  };

  public updateSaleTerms = async (
    propertyId: number,
    saleTermId: number,
    saleTerms: IUpdateSaleTermDetails
  ): Promise<void> => {
    return await this.apiClient.put(ENDPOINTS.updateSaleTerms(propertyId, saleTermId), saleTerms);
  };

  public getExistingVerificationDocuments = async (propertyId: number): Promise<IVerificationDocumentList[]> => {
    return await this.apiClient.get(ENDPOINTS.existingVerificationDocuments(propertyId));
  };

  public deleteVerificationDocument = async (propertyId: number, documentId: number): Promise<void> => {
    return await this.apiClient.delete(ENDPOINTS.deleteExistingVerificationDocuments(propertyId, documentId));
  };

  public getPropertyImagesByPropertyId = async (propertyId: number): Promise<IPropertySelectedImages[]> => {
    return await this.apiClient.get(ENDPOINTS.assetAttachments(propertyId));
  };

  public markAttachmentAsCoverImage = async (
    propertyId: number,
    attachmentId: number
  ): Promise<IMarkCoverImageAttachment> => {
    return await this.apiClient.put(ENDPOINTS.markAttachmentAsCoverImage(propertyId, attachmentId));
  };

  public postAttachmentsForProperty = async (
    propertyId: number,
    requestBody: IPropertyImagesPostPayload[]
  ): Promise<void> => {
    return await this.apiClient.post(ENDPOINTS.assetAttachments(propertyId), requestBody);
  };

  public postVerificationDocuments = async (
    propertyId: number,
    requestBody: IPostVerificationDocuments[]
  ): Promise<void> => {
    return await this.apiClient.post(ENDPOINTS.postVerificationDocuments(propertyId), requestBody);
  };

  public deletePropertyImage = async (attachmentId: number): Promise<void> => {
    return await this.apiClient.delete(ENDPOINTS.deletePropertyAttachment(attachmentId));
  };

  public getAssetIdentityDocuments = async (): Promise<IVerificationDocumentList[]> => {
    return await this.apiClient.get(ENDPOINTS.assetIdentityDocuments());
  };

  public getVerificationDocumentTypes = async (): Promise<IVerificationTypes[]> => {
    return await this.apiClient.get(ENDPOINTS.getVerificationDocumentDetails());
  };

  public getRatings = async (id: number): Promise<AssetReview[]> => {
    const response = await this.apiClient.get(ENDPOINTS.getRatings(id));
    return ObjectMapper.deserializeArray(AssetReview, response);
  };

  public getLeaseListing = async (propertyTermId: number): Promise<Asset> => {
    const response = await this.apiClient.get(ENDPOINTS.getLeaseListing(propertyTermId));
    return ObjectMapper.deserialize(Asset, response);
  };

  public getSaleListing = async (propertyTermId: number): Promise<Asset> => {
    const response = await this.apiClient.get(ENDPOINTS.getSaleListing(propertyTermId));
    return ObjectMapper.deserialize(Asset, response);
  };

  public getSimilarProperties = async (propertyTermId: number, type: number): Promise<Asset[]> => {
    if (type === 0) {
      // RENT FLOW
      const response = await this.apiClient.get(ENDPOINTS.getSimilarPropertiesForLease(propertyTermId));
      return ObjectMapper.deserializeArray(Asset, response);
    }
    // SALE FLOW
    const response = await this.apiClient.get(ENDPOINTS.getSimilarPropertiesForSale(propertyTermId));
    return ObjectMapper.deserializeArray(Asset, response);
  };
}

const propertyRepository = new AssetRepository();
export { propertyRepository as AssetRepository };
