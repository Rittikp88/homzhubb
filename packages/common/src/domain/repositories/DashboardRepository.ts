import { BootstrapAppService } from '@homzhub/common/src/services/BootstrapAppService';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { IApiClient } from '@homzhub/common/src/network/Interfaces';
import { MarketTrends } from '@homzhub/common/src/domain/models/MarketTrends';
import { AssetMetrics } from '@homzhub/common/src/domain/models/AssetMetrics';
import { AssetAdvertisement } from '@homzhub/common/src/domain/models/AssetAdvertisement';
import { GeneralLedgers } from '@homzhub/common/src/domain/models/GeneralLedgers';

const ENDPOINTS = {
  getMarketTrends: (): string => 'market-trends/',
  getAssetMetrics: (): string => 'dashboard/management-tab/',
  getAdvertisements: (): string => 'advertisements/',
  getGeneralLedgers: (): string => 'general-ledgers/overall-performances/',
};

class DashboardRepository {
  private apiClient: IApiClient;

  public constructor() {
    this.apiClient = BootstrapAppService.clientInstance;
  }

  public getMarketTrends = async (limit: number): Promise<MarketTrends> => {
    const response = await this.apiClient.get(ENDPOINTS.getMarketTrends(), { limit });
    return ObjectMapper.deserialize(MarketTrends, response);
  };

  public getAssetMetrics = async (): Promise<AssetMetrics> => {
    const response = await this.apiClient.get(ENDPOINTS.getAssetMetrics());
    return ObjectMapper.deserialize(AssetMetrics, response);
  };

  public getAdvertisements = async (): Promise<AssetAdvertisement> => {
    const response = await this.apiClient.get(ENDPOINTS.getAdvertisements());
    return ObjectMapper.deserialize(AssetAdvertisement, response);
  };

  public getGeneralLedgers = async (requestPayload: {
    transaction_date__lte: string;
    transaction_date__gte: string;
    transaction_date_group_by: string;
  }): Promise<GeneralLedgers[]> => {
    const response = await this.apiClient.get(ENDPOINTS.getGeneralLedgers(), requestPayload);
    return ObjectMapper.deserializeArray(GeneralLedgers, response);
  };
}

const dashboardRepository = new DashboardRepository();
export { dashboardRepository as DashboardRepository };
