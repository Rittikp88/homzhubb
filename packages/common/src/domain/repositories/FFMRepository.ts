import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { BootstrapAppService } from '@homzhub/common/src/services/BootstrapAppService';
import { Feedback } from '@homzhub/common/src/domain/models/Feedback';
import { FFMVisit } from '@homzhub/common/src/domain/models/FFMVisit';
import { InspectionReport } from '@homzhub/common/src/domain/models/InspectionReport';
import { OnBoarding } from '@homzhub/common/src/domain/models/OnBoarding';
import { OutsetCheck } from '@homzhub/common/src/domain/models/OutsetCheck';
import { Unit } from '@homzhub/common/src/domain/models/Unit';
import { IApiClient } from '@homzhub/common/src/network/Interfaces';
import {
  IFFMVisitParam,
  IGetFeedbackParam,
  IOutsetCheckParam,
  IPostFeedback,
  IUpdateReport,
} from '@homzhub/common/src/domain/repositories/interfaces';

const ENDPOINTS = {
  onBoarding: 'v1/ffm-onboardings',
  roles: 'v1/roles',
  visits: 'v1/ffm/tasks/site-visits/',
  visitDetail: (visitId: number): string => `v1/ffm/tasks/site-visits/${visitId}/`,
  rejectReason: (visitId: number): string => `v1/ffm/listing-visits/${visitId}/prospect-feedbacks/reject-reasons/`,
  feedback: (visitId: number): string => `v1/ffm/listing-visits/${visitId}/prospect-feedbacks/`,
  feedbackById: (visitId: number, feedbackId: number): string =>
    `v1/ffm/listing-visits/${visitId}/prospect-feedbacks/${feedbackId}`,
  inspectionReport: 'v1/ffm/tasks/inspection-reports/',
  inspectionReportById: (reportId: number): string => `v1/ffm/tasks/inspection-reports/${reportId}/`,
  outsetsCheck: (reportId: number): string => `v1/ffm/tasks/inspection-reports/${reportId}/outsets/`,
};

class FFMRepository {
  private apiClient: IApiClient;

  public constructor() {
    this.apiClient = BootstrapAppService.clientInstance;
  }

  public getOnBoarding = async (): Promise<OnBoarding[]> => {
    const response = await this.apiClient.get(ENDPOINTS.onBoarding);
    return ObjectMapper.deserializeArray(OnBoarding, response);
  };

  public getRoles = async (): Promise<Unit[]> => {
    const response = await this.apiClient.get(ENDPOINTS.roles);
    return ObjectMapper.deserializeArray(Unit, response);
  };

  public getVisits = async (param?: IFFMVisitParam): Promise<FFMVisit[]> => {
    const response = await this.apiClient.get(ENDPOINTS.visits, param);
    return ObjectMapper.deserializeArray(FFMVisit, response);
  };

  public getVisitDetail = async (visitId: number): Promise<FFMVisit> => {
    const response = await this.apiClient.get(ENDPOINTS.visitDetail(visitId));
    return ObjectMapper.deserialize(FFMVisit, response);
  };

  public getRejectReason = async (visitId: number): Promise<Unit[]> => {
    const response = await this.apiClient.get(ENDPOINTS.rejectReason(visitId));
    return ObjectMapper.deserializeArray(Unit, response);
  };

  public postFeedback = async (payload: IPostFeedback): Promise<void> => {
    const { visitId, data } = payload;
    return await this.apiClient.post(ENDPOINTS.feedback(visitId), data);
  };

  public getFeedbackById = async (payload: IGetFeedbackParam): Promise<Feedback> => {
    const response = await this.apiClient.get(ENDPOINTS.feedbackById(payload.visitId, payload.feedbackId));
    return ObjectMapper.deserialize(Feedback, response);
  };

  public getInspectionReport = async (payload: string): Promise<InspectionReport> => {
    const response = await this.apiClient.get(ENDPOINTS.inspectionReport, { status_category: payload });
    return ObjectMapper.deserialize(InspectionReport, response);
  };

  public updateInspectionReport = async (payload: IUpdateReport): Promise<void> => {
    return await this.apiClient.patch(ENDPOINTS.inspectionReportById(payload.reportId), {
      status: payload.status,
    });
  };

  public outsetsCheck = async (payload: IOutsetCheckParam): Promise<OutsetCheck> => {
    const response = await this.apiClient.post(ENDPOINTS.outsetsCheck(payload.reportId), payload.body);
    return ObjectMapper.deserialize(OutsetCheck, response);
  };
}

const ffmRepository = new FFMRepository();
export { ffmRepository as FFMRepository };
