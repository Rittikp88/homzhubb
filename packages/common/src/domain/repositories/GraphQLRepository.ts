import axios from 'axios';

const ENDPOINTS = {
  baseUrl: 'https://api-ap-northeast-1.graphcms.com/v2/ckiqymr7djhfb01z1di6m4fjs/master',
};

export interface IFeaturedProperties {
  address: string;
  category: string;
  coverImage: { url: string };
  developer: { name: string };
  featured: boolean;
  id: string;
  images: { url: string }[];
  possessionDate: string;
  priceRange: string;
  projectName: string;
  slug: string;
  stage: string;
  typesAvailable: string;
}

const Queries = {
  featuredProperties: `query {
        properties(orderBy: updatedAt_ASC) {
            id
            address
            possessionDate
            category
            priceRange
            typesAvailable
            coverImage {
                url
            }
            developer {
                name
            }
            projectName
            slug
            stage
            featured
            images {
                url
            }
        }
    }`,
};
class GraphQLRepository {
  public getFeaturedProperties = async (): Promise<IFeaturedProperties[]> => {
    const { featuredProperties } = Queries;
    const response = await axios({
      url: ENDPOINTS.baseUrl,
      method: 'post',
      data: {
        query: featuredProperties,
      },
    }).then((result) => {
      return result.data;
    });
    return response.data.properties;
  };
}

const graphQLRepository = new GraphQLRepository();
export { graphQLRepository as GraphQLRepository };
