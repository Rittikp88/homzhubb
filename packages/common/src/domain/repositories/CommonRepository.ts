import { BootstrapAppService } from '@homzhub/common/src/services/BootstrapAppService';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { CarpetArea } from '@homzhub/common/src/domain/models/CarpetArea';
import { Country } from '@homzhub/common/src/domain/models/Country';
import { Currency } from '@homzhub/common/src/domain/models/Currency';
import { MarketTrends } from '@homzhub/common/src/domain/models/MarketTrends';
import { OnBoarding } from '@homzhub/common/src/domain/models/OnBoarding';
import { SocialAuthProvider } from '@homzhub/common/src/domain/models/SocialAuthProvider';
import { Unit } from '@homzhub/common/src/domain/models/Unit';
import { User } from '@homzhub/common/src/domain/models/User';
import { IApiClient } from '@homzhub/common/src/network/Interfaces';
import { ISupportPayload, IMarketTrendParams } from '@homzhub/common/src/domain/repositories/interfaces';
import { SocialAuthProviders } from '@homzhub/common/src/constants/SocialAuthProviders';

const ENDPOINTS = {
  getCountryCodes: 'countries/',
  getCurrencyCodes: 'currency-codes/',
  carpetAreaUnits: 'carpet-area-units/',
  maintenanceUnits: 'list-of-values/maintenance-units/',
  onBoarding: 'onboardings/',
  supportCategories: 'list-of-values/client-support-categories/',
  supportContact: 'client-support/contacts/',
  clientSupport: 'client-support/',
  getMarketTrends: 'market-trends/',
  updateMarketTrends: (id: number): string => `market-trends/${id}/`,
};

class CommonRepository {
  private apiClient: IApiClient;

  public constructor() {
    this.apiClient = BootstrapAppService.clientInstance;
  }

  public getCountryCodes = async (): Promise<Country[]> => {
    const response = await this.apiClient.get(ENDPOINTS.getCountryCodes);
    return ObjectMapper.deserializeArray(Country, response);
  };

  public getCurrencyCodes = async (): Promise<Currency[]> => {
    const response = await this.apiClient.get(ENDPOINTS.getCurrencyCodes);
    return ObjectMapper.deserializeArray(Currency, response);
  };

  public getCarpetAreaUnits = async (): Promise<CarpetArea[]> => {
    const response = await this.apiClient.get(ENDPOINTS.carpetAreaUnits);
    return ObjectMapper.deserializeArray(CarpetArea, response);
  };

  public getOnBoarding = async (): Promise<OnBoarding[]> => {
    const response = await this.apiClient.get(ENDPOINTS.onBoarding);
    return ObjectMapper.deserializeArray(OnBoarding, response);
  };

  public getSocialMedia = (): SocialAuthProvider[] => {
    return ObjectMapper.deserializeArray(SocialAuthProvider, SocialAuthProviders);
  };

  public getMaintenanceUnits = async (): Promise<Unit[]> => {
    const response = await this.apiClient.get(ENDPOINTS.maintenanceUnits);
    return ObjectMapper.deserializeArray(Unit, response);
  };

  public getSupportCategories = async (): Promise<Unit[]> => {
    const response = await this.apiClient.get(ENDPOINTS.supportCategories);
    return ObjectMapper.deserializeArray(Unit, response);
  };

  public getSupportContacts = async (): Promise<User> => {
    const response = await this.apiClient.get(ENDPOINTS.supportContact);
    return ObjectMapper.deserialize(User, response);
  };

  public postClientSupport = async (payload: ISupportPayload): Promise<void> => {
    return await this.apiClient.post(ENDPOINTS.clientSupport, payload);
  };

  public getMarketTrends = async (params: IMarketTrendParams): Promise<MarketTrends> => {
    const response = await this.apiClient.get(ENDPOINTS.getMarketTrends, params);
    return ObjectMapper.deserialize(MarketTrends, response);
  };

  public updateMarketTrends = async (id: number): Promise<void> => {
    await this.apiClient.patch(ENDPOINTS.updateMarketTrends(id));
  };
}

const commonRepository = new CommonRepository();
export { commonRepository as CommonRepository };
