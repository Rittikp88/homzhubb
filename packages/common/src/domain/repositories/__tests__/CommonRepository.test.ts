import { BootstrapAppService } from '@homzhub/common/src/services/BootstrapAppService';
import { CommonRepository } from '@homzhub/common/src/domain/repositories/CommonRepository';
import { countryCodes, carpetAreaUnits, currencyCodes } from '@homzhub/common/src/mocks/CommonRepositoryMocks';
import { OnboardingData } from '@homzhub/common/src/mocks/onboarding';
import { SocialMediaData } from '@homzhub/common/src/mocks/socialMedia';

jest.mock('@homzhub/common/src/services/storage/StorageService', () => 'StorageService');
jest.mock('@react-native-community/google-signin', () => {});

describe('CommonRepository', () => {
  afterAll(() => {
    jest.resetAllMocks();
  });

  it('should fetch a list of county codes', async () => {
    // @ts-ignore
    jest.spyOn(BootstrapAppService.clientInstance, 'get').mockImplementation(() => countryCodes);
    const response = await CommonRepository.getCountryCodes();
    expect(response).toMatchSnapshot();
  });

  it('should fetch a list of unit areas', async () => {
    // @ts-ignore
    jest.spyOn(BootstrapAppService.clientInstance, 'get').mockImplementation(() => carpetAreaUnits);
    const response = await CommonRepository.getCarpetAreaUnits();
    expect(response).toMatchSnapshot();
  });

  it('should fetch a list of currency codes', async () => {
    // @ts-ignore
    jest.spyOn(BootstrapAppService.clientInstance, 'get').mockImplementation(() => currencyCodes);
    const response = await CommonRepository.getCountryCodes();
    expect(response).toMatchSnapshot();
  });

  it('should fetch social media data', async () => {
    // @ts-ignore
    jest.spyOn(BootstrapAppService.clientInstance, 'get').mockImplementation(() => SocialMediaData);
    const response = await CommonRepository.getSocialMedia();
    expect(response).toMatchSnapshot();
  });

  it('should fetch OnBoarding screen data', async () => {
    // @ts-ignore
    jest.spyOn(BootstrapAppService.clientInstance, 'get').mockImplementation(() => OnboardingData);
    const response = await CommonRepository.getOnboarding();
    expect(response).toMatchSnapshot();
  });

  ['getCountryCodes', 'getCarpetAreaUnits', 'getOnboarding', 'getCurrencyCodes', 'getSocialMedia'].forEach(
    (api: string) => {
      it(`should throw an error in case of ${api} API failure`, async () => {
        jest.spyOn(BootstrapAppService.clientInstance, 'get').mockImplementation(() => {
          throw new Error('Test Error');
        });
        try {
          // @ts-ignore
          await CommonRepository[api]();
        } catch (e) {
          expect(e).toBeTruthy();
        }
      });
    }
  );
});
