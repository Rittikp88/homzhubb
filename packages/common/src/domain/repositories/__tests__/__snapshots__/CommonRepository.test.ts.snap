// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonRepository should fetch OnBoarding screen data 1`] = `
Array [
  Object {
    "description": "Sit amet, in enim nunc, etiam amet velit. Malesuada non neque nec in a.",
    "image_url": "https://homzhub-bucket.s3.amazonaws.com/Search_Onboarding.svg",
    "title": "Search",
  },
  Object {
    "description": "Sit amet, in enim nunc, etiam amet velit. Malesuada non neque nec in a.",
    "image_url": "https://homzhub-bucket.s3.amazonaws.com/Post_Onboarding.svg",
    "title": "Post",
  },
  Object {
    "description": "Sit amet, in enim nunc, etiam amet velit. Malesuada non neque nec in a.",
    "image_url": "https://homzhub-bucket.s3.amazonaws.com/Maintain_Onboarding.svg",
    "title": "Maintain",
  },
]
`;

exports[`CommonRepository should fetch a list of county codes 1`] = `
Array [
  Object {
    "alpha_code": "IN",
    "country": "INDIA",
    "country_code": "+91",
    "id": 1,
  },
  Object {
    "alpha_code": "SG",
    "country": "SINGAPORE",
    "country_code": "+65",
    "id": 2,
  },
  Object {
    "alpha_code": "AU",
    "country": "AUSTRALIA",
    "country_code": "+36",
    "id": 3,
  },
]
`;

exports[`CommonRepository should fetch a list of currency codes 1`] = `
Array [
  Object {
    "currency_code": "AUD",
    "currency_symbol": "$",
  },
  Object {
    "currency_code": "INR",
    "currency_symbol": "â‚¹",
  },
  Object {
    "currency_code": "SGD",
    "currency_symbol": "$",
  },
]
`;

exports[`CommonRepository should fetch a list of unit areas 1`] = `
Array [
  Object {
    "label": "Sq.ft",
    "value": "SQ_FT",
  },
  Object {
    "label": "Sq.yard",
    "value": "SQ_YARD",
  },
  Object {
    "label": "Sq.m",
    "value": "SQ_M",
  },
  Object {
    "label": "Cents",
    "value": "CENTS",
  },
  Object {
    "label": "Biswa",
    "value": "BISWA",
  },
  Object {
    "label": "Acres",
    "value": "ACRES",
  },
  Object {
    "label": "Hectares",
    "value": "HECTARES",
  },
  Object {
    "label": "Kattha",
    "value": "KATTHA",
  },
]
`;

exports[`CommonRepository should fetch social media data 1`] = `
Array [
  Object {
    "clientID": "1008014324415-kclkc41epc41nqjp0uc4gdgg528m1nc6.apps.googleusercontent.com",
    "description": "with Google",
    "provider": "GOOGLE",
  },
  Object {
    "clientID": "",
    "description": "with Facebook",
    "provider": "FACEBOOK",
  },
  Object {
    "clientID": "",
    "description": "with Linkedin",
    "provider": "LINKEDIN",
  },
]
`;
